File: plugins/discourse-workflows/lib/discourse_workflows/node_issues.rb
# frozen_string_literal: true

module DiscourseWorkflows
  class NodeIssues
    class << self
      def for_node(node, node_type_class)
        return [] unless node_type_class.respond_to?(:property_schema)
        schema = node_type_class.property_schema || {}
        config = NodeData.resolved_parameters(node)
        walk(schema, config.deep_stringify_keys, "")
      end

      private

      def walk(schema, config, path_prefix)
        effective = apply_defaults(schema, config)
        issues = []

        schema.each do |name, field|
          next unless field_definitely_visible?(field, effective)

          name_s = name.to_s
          path = path_prefix.empty? ? name_s : "#{path_prefix}.#{name_s}"
          value = effective[name_s]

          if field[:required] && blank_value?(value)
            issues << { path: path, name: name_s, message: "required" }
          end

          case field[:type]
          when :collection
            if value.is_a?(Hash)
              field[:options].to_a.each do |option|
                option_name = option[:name].to_s
                next unless value.key?(option_name)

                option_schema = { option_name => option.except(:name, :display_name) }
                issues.concat(walk(option_schema, value, path))
              end
            elsif value.is_a?(Array)
              item_schema = (field[:item_schema] || {}).merge(field[:extra_item_schema] || {})
              value.each_with_index do |item, index|
                item_hash = item.is_a?(Hash) ? item.deep_stringify_keys : {}
                issues.concat(walk(item_schema, item_hash, "#{path}.#{index}"))
              end
            end
          when :fixed_collection
            field[:options].to_a.each do |group|
              group_name = group[:name].to_s
              rows = CollectionParameters.rows_from_value(value, group: group_name)
              if field[:required] && !blank_value?(value) && blank_value?(rows)
                issues << { path: "#{path}.#{group_name}", name: name_s, message: "required" }
              end

              rows.each_with_index do |item, index|
                item_hash = item.is_a?(Hash) ? item.deep_stringify_keys : {}
                issues.concat(
                  walk(group[:values] || {}, item_hash, "#{path}.#{group_name}.#{index}"),
                )
              end
            end
          when :assignment_collection
            assignments = CollectionParameters.rows_from_value(value, group: "assignments")
            if field[:required] && !blank_value?(value) && blank_value?(assignments)
              issues << { path: "#{path}.assignments", name: name_s, message: "required" }
            end

            assignment_schema = {
              name: {
                type: :string,
                required: true,
              },
              type: {
                type: :options,
                required: true,
              },
              value: {
                type: :string,
                required: true,
              },
            }

            assignments.each_with_index do |item, index|
              item_hash = item.is_a?(Hash) ? item.deep_stringify_keys : {}
              issues.concat(walk(assignment_schema, item_hash, "#{path}.assignments.#{index}"))
            end
          end
        end
        issues
      end

      def apply_defaults(schema, config)
        result = config.dup
        schema.each do |name, field|
          name_s = name.to_s
          result[name_s] = field[:default] if result[name_s].nil? && !field[:default].nil?
        end
        result
      end

      def field_definitely_visible?(field, config)
        return false if (field[:ui] || {})[:hidden]

        Schema.definitely_visible?(field[:display_options] || {}, config)
      end

      def blank_value?(value)
        return true if value.nil?
        return value.strip.empty? if value.is_a?(String)
        return value.empty? if value.is_a?(Array)
        false
      end
    end
  end
end

File: plugins/discourse-workflows/admin/assets/javascripts/admin/lib/workflows/node-issues.js
import {
  fieldType,
  fieldVisible,
  fixedCollectionGroups,
  fixedCollectionRows,
  normalizePropertyOptions,
} from "./property-engine";

function isBlank(value) {
  if (value === undefined || value === null) {
    return true;
  }
  if (typeof value === "string") {
    return value.trim() === "";
  }
  if (Array.isArray(value)) {
    return value.length === 0;
  }
  return false;
}

function itemSchemaFor(field) {
  return {
    ...(field.item_schema || {}),
    ...(field.extra_item_schema || {}),
  };
}

function fixedCollectionItemSchema(group) {
  return group.values || {};
}

function applyDefaults(schema, config) {
  const effective = { ...(config || {}) };
  for (const [name, field] of Object.entries(schema)) {
    if (effective[name] == null && field.default !== undefined) {
      effective[name] = field.default;
    }
  }
  return effective;
}

function walk(schema, config, pathPrefix, issues) {
  const effective = applyDefaults(schema, config);

  for (const [name, field] of Object.entries(schema)) {
    if (!fieldVisible(field, effective)) {
      continue;
    }

    const path = pathPrefix ? `${pathPrefix}.${name}` : name;
    const value = effective[name];

    if (field.required && isBlank(value)) {
      issues.push({ path, name, message: "required" });
    }

    const type = fieldType(field);

    if (
      type === "collection" &&
      value &&
      typeof value === "object" &&
      !Array.isArray(value)
    ) {
      normalizePropertyOptions(field.options || []).forEach((option) => {
        if (Object.hasOwn(value, option.name)) {
          walk({ [option.name]: option }, value, path, issues);
        }
      });
    } else if (
      type === "fixed_collection" &&
      value &&
      typeof value === "object"
    ) {
      fixedCollectionGroups(field).forEach((group) => {
        const rows = fixedCollectionRows(value, group.name);
        if (field.required && isBlank(rows)) {
          issues.push({
            path: `${path}.${group.name}`,
            name,
            message: "required",
          });
        }
        rows.forEach((item, index) => {
          walk(
            fixedCollectionItemSchema(group),
            item || {},
            `${path}.${group.name}.${index}`,
            issues
          );
        });
      });
    } else if (
      type === "assignment_collection" &&
      value &&
      typeof value === "object"
    ) {
      const assignments = fixedCollectionRows(value, "assignments");
      if (field.required && isBlank(assignments)) {
        issues.push({ path: `${path}.assignments`, name, message: "required" });
      }
      assignments.forEach((item, index) => {
        walk(
          {
            name: { type: "string", required: true },
            type: { type: "options", required: true },
            value: { type: "string", required: true },
          },
          item || {},
          `${path}.assignments.${index}`,
          issues
        );
      });
    } else if (field.type === "collection" && Array.isArray(value)) {
      const inner = itemSchemaFor(field);
      value.forEach((item, index) => {
        walk(inner, item || {}, `${path}.${index}`, issues);
      });
    }
  }
}

export default function getNodeIssues(configuration, propertySchema) {
  if (!propertySchema) {
    return [];
  }
  const issues = [];
  walk(propertySchema, configuration || {}, "", issues);
  return issues;
}
